Introduce common structure for key counters

ItsShamed opened the pull request:
Supersedes #22640

I decided to split the abstraction and the new design implementation as suggested in https://github.com/ppy/osu/pull/22640#issuecomment-1430778480

This PR abstracts the `KeyCounter` and `KeyCounterDisplay` in order to simplfy the implementation of more designs.
This also simplifies the creation of more input triggers for the `KeyCounter`s to react to.

Some commits have extended messages, in case the diff is too messy to understand anything.

Most of these commits were just cherry-picked from #22640

ItsShamed committed:
refactor: separate things in KeyCounter

To implement different different sources of input for KeyCounter, it
is now possible to create a Trigger class (to inherit) instead of
inheriting KeyCounter. This eases the creation of more input sources
(like for tests) while allowing to implement different UI variants.

That way, if another variant of the key counter needs to implemented
(for whathever reason), this can be done by only inheriting KeyCounter
and changing how things are arranged visually.

ItsShamed committed:
refactor: separate impl of KeyCounterDisplay

This allows for different layouts of display. Idk, maybe someone would
want to mix both variants? (don't do this please). This commit is mostly
prep for further changes.

ItsShamed committed:
refactor: hide trigger presence from content

ItsShamed committed:
refactor: rename `Trigger` class to `InputTrigger`

ItsShamed committed:
fix(KeyCounter): don't override Handle

This caused the Keyboard inputs to register twice, which is not what we
want.

ItsShamed committed:
style(KeyCounter): rename `(Un)lit` methods to `(Un)light`

ItsShamed committed:
refactor: rename trigger classes

Makes it better to understand their purpose

ItsShamed committed:
style: nullable pass on `KeyCounterDisplay`

ItsShamed committed:
reafactor: simplify type checking

peppy commented:
I'm not too sure I understand why you've added non-action based key triggers. Before I begin to review, could you explain that?

ItsShamed commented:
miss click was about to answer smh

ItsShamed commented:
> I'm not too sure I understand why you've added non-action based key triggers. Before I begin to review, could you explain that?

If you're talking about `KeyCounterKeyboardTrigger` and `KeyCounterMouseTrigger` they basically were `KeyCounter`s themselves before the PR. So in this PR, instead of being `KeyCounter`s, they are now Triggers. The only place they were used is in the tests:

- master

https://github.com/ppy/osu/blob/87901bc8b1b1f5c1acb215697040c99905063ba1/osu.Game.Tests/Visual/Gameplay/TestSceneKeyCounter.cs#L28-L31

- this PR

https://github.com/ppy/osu/blob/74e7cc205601bc4edb9d88e12afb8c63f8e967d2/osu.Game.Tests/Visual/Gameplay/TestSceneKeyCounter.cs#L34-L37

Some other tests also had custom `KeyCounter`s for custom actions, so they were also converted to Triggers.

I hope I was clear enough, even though I thought this would be obvious to think about at first

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -69,82 +67,36 @@ public void Decrement()
-        private void load(TextureStore textures)
-        {
-            Children = new Drawable[]
+            public KeyCounter Target
```
The circularity of this is giving me a headache. `KeyCounter`s have instances of `InputTrigger`, which are exposed so that the `RulesetInputManager` can access the `InputTrigger`s and call methods on them, which... end up performing actions on the `KeyCounter` again?

I'd like to see looser coupling here. Maybe have `InputTrigger` only expose events and be blissfully unaware of the key counter, and the counter can subscribe to the events and do what it needs to do.

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -1,58 +1,56 @@
 
-        public bool IsCounting { get; set; } = true;
-        private int countPresses;
+        protected Bindable<bool> IsCountingBindable = new BindableBool(true);
```
Why is this a thing? It's protected, not accessed by any subclass, and it's only used in `IsCounting` which is... a plain bool property?

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -1,58 +1,56 @@
         }
 
-        private bool isLit;
+        public int CountPresses
+        {
+            get => PressesCount.Value;
+            private set => PressesCount.Value = value;
+        }
```
This is also really weird; the naming variances aside, it's essentially a getter for `PressesCount` and I'm not sure why the setter exists if it's private.

I'd say just expose the bindable. If you're worried about outward mutation, then do this:

```csharp
private BindableInt count = new BindableInt();
public IBindable<int> Count => count;
```

Consumers can still `.BindValueChanged()` on `Count` but not modify it.

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -1,58 +1,56 @@
+            Name = trigger.Name;
         }
 
+        protected Bindable<bool> IsLit = new BindableBool();
```
I'd depart from the "lighting:" vernacular here. I'm not sure it's going to make sense for all implementations ever. My proposal:

- `IsLit` -> `IsActive`
- `Light()` -> `Activate()`
- `Unlight()` -> `Deactivate()`

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -69,82 +67,36 @@ public void Decrement()
+                Name = name;
+            }
+
+            protected void Light(bool increment = true)
```
My favourites, bool arguments, and opaque ones at that. At face value nobody is going to know as to whether they want to "increment" or "preserve" or whatever. As far as I gather this is essentially for rewind? Then I'd call it by name:

```csharp
protected void Activate(bool forwardPlayback = true);
```

or even better yet

```csharp
protected void Activate(int playbackDirection);
```

and then you do `Activate(1)` when going forwards and `Activate(-1)` when doing backwards.

peppy replied:
The worst part is the `Unlight` method which has a different named argument which is doing the ... same thing in reverse? My brain hurt.

bdach replied:
@ItsShamed This is not fully addressed; while `InputTrigger` looks better wrt naming, `Activate(bool increment)` and `Deactivate(bool preserve)` are still there.

ItsShamed replied:
Oh yeah, my brain trolled me :trollface: 
I thought I did a full pass in c61fac578ca53fa4dba22ac7ec85aa0cc335c762

bdach commented on osu.Game/Screens/Play/KeyCounterDisplay.cs:
```diff
@@ -12,18 +10,14 @@
+        protected readonly Bindable<bool> ConfigVisibility = new Bindable<bool>();
 
-        protected readonly FillFlowContainer<KeyCounter> KeyFlow;
+        protected FillFlowContainer<KeyCounter> KeyFlow = new FillFlowContainer<KeyCounter>();
```
This is settable by subclasses now - I'd hope that's not intentional?

bdach commented on osu.Game/Screens/Play/KeyCounterDisplay.cs:
```diff
@@ -33,48 +27,29 @@ public partial class KeyCounterDisplay : Container<KeyCounter>
-            key.KeyUpTextColor = KeyUpTextColor;
         }
 
+        protected virtual bool CheckType(KeyCounter key) => true;
```
This is very weird. What is this for? As best I can tell it's to avoid mixing up key counters of different drawable implementations?

Hesitantly I'd suggest to make the class a generic (as in `KeyCounterDisplay<TKeyCounter> where TKeyCounter : KeyCounter`) or something. I'm not sure it's ever going to be valid to mix these?

ItsShamed replied:
I don't know what to do for this one, as I feel like there could be a lot of directions, but I don't know which one is the best.

From what I can learn from other generics in the codebase, I could separate the class into generic and non-generic base classes. 
However, forcing the new implementations to inherit the generic base only is pretty _impossible_ for this class, as making the constructor `internal` on the non-generic has pretty much no effect since new implementations will essentially be created in the same assembly.

I could also just make the base class generic without splitting anything, but then it would mean that every occurrence of `KeyCounterDisplay` would have to be replaced with `KeyCounterDisplay<KeyCounter>`, and I don't know if it's okay to do that (especially for variable declarations).

peppy replied:
Why does the check even need to exist?

bdach reviewed (commented):
First pass. Mostly examination of structure.

ItsShamed committed:
refactor(KeyCounter): remove circularity

ItsShamed committed:
refactor(KeyCounter): address bindables issues

`IsCounting` is back being an auto-property.
`countPresses` is now encapsulated and being exposed as an
`IBindable<int>` via `CountPresses`

ItsShamed committed:
style(KeyCounter): rename methods and arguments

As for the second suggestion in
https://github.com/ppy/osu/pull/22654#discussion_r1109047998,
I went with the first one as only one Trigger actually uses this
argument for rewinding.

ItsShamed committed:
refactor: make `FillFlowContainer` read-only

ItsShamed committed:
fix(TestSceneOsuTouchInput): missing Value call

ItsShamed committed:
style(KeyCounter): rename `IsLit` to `IsActive`

ItsShamed committed:
style(KeyCounterDisplay): remove type check

ItsShamed committed:
fix: compilation

ItsShamed commented:
All issues have been addressed (hopefully).

peppy commented on osu.Game/Rulesets/UI/RulesetInputManager.cs:
```diff
@@ -166,7 +166,7 @@ public void Attach(KeyCounterDisplay keyCounter)
                                                    .Distinct()
                                                    .OrderBy(action => action)
-                                                   .Select(action => new KeyCounterAction<T>(action)));
+                                                   .Select(action => keyCounter.CreateKeyCounter(new KeyCounterActionTrigger<T>(action))));
```
Sorry but I'm just not feeling this. It all seems back to front.

Whenever you see a class passing something to itself *created by a method on that class* you know something's up.

This needs to either:

- Change back to `keyCounter.Add(new KeyCounterAction<T>` (what it was before), moving actual implementation differences out of that class and making it a model class (ie. non-drawable).
- Change to `keyCounter.AddCounter<T>(action)`.

@bdach could you please confirm you agree that the way this PR is currently heading feels wrong?

bdach replied:
Agree that this is too weird to live.

Either resolution sounds fine, but I will note that in general I'm really not sure how long `KeyCounterDisplay` will live in the state proposed in this PR wherein it's just a fill flow with separate `KeyCounter`s. This is going to make a lot of stuff difficult to do, including for instance the legacy key counter. To me the key counter display (as a whole, all of its pieces) should be a single interchangeable skinnable and the simple flow layout should be moved out to implementations. Which I guess is probably closer to option 1 in general direction.

ItsShamed replied:
So this means that there will be a single `KeyCounterDisplay` and we're moving back to the old `KeyCounter` inheritance scheme, is that right? 

If that's the case, then how to provide the individual _display implementations_?

I'm not quite sure if I fully understood your directions...

bdach replied:
Very fast and loose sketch of how I imagine this working:

```csharp
public abstract class KeyCounterDisplay : CompositeDrawable
{
	public readonly Bindable<bool> AlwaysVisible { get; }
	public readonly bool IsCounting { get; set; } // maybe consider bindable after all?
	
	protected readonly Bindable<bool> ConfigVisibility;
	
	protected abstract void UpdateVisibility();
	
	public void SetReceptor(Receptor receptor);
	public void AddTrigger(KeyCounter.InputTrigger trigger); // probably move out InputTrigger so that it's not an inner class anymore
	
	protected abstract KeyCounter CreateKeyCounter(KeyCounter.InputTrigger trigger);
	
	// notably, Content, KeyFlow, etc. are not here - this class has no layout logic, it just hooks up data flows
}

public class ArgonKeyCounterDisplay : KeyCounterDisplay { // ... }
public class TrianglesKeyCounterDisplay : KeyCounterDisplay { // ... }
public class LegacyKeyCounterDisplay : KeyCounterDisplay { // ... }
```

Would like @peppy's opinion before you act on it though.

ItsShamed replied:
I thought at some point of doing something like this, but I guess I overengineered things in my head which led me to put this idea aside.

peppy replied:
This might be okay, so long as `new InputTrigger()` is possible.

ItsShamed replied:
Do you mean that I should make `InputTrigger` not abstract with a public parameterless constructor?

peppy replied:
It can be abstract as long as it's not per-skin.

ItsShamed replied:
Oh okay. Well, it has never been the goal to make it skin dependent and will never be, they are just only used to trigger Activation and Deactivation states for the `KeyCounter` to react to. 
Also see the recent changes, I've applied the suggestions from bdach.

ItsShamed committed:
style(KeyCounter): `forwardPlayback`

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -1,151 +1,99 @@
 
-            CountPresses--;
+            countPresses.Value--;
         }
```
Not sure these should be public now. Only callers are the protected event flows.

bdach commented on osu.Game/Screens/Play/KeyCounter.cs:
```diff
@@ -1,151 +1,99 @@
+            Name = trigger.Name;
         }
 
+        protected Bindable<bool> IsActive = new BindableBool();
```
Should be `private set` or `readonly`

ItsShamed committed:
style(KeyCounter): fields and methods visiblity

ItsShamed committed:
refactor(KeyCounterDisplay): apply suggestions

I also took the freedom to add type checking, as we can't limit the
usage of `Add()` since it's a Container. The exception thrown also
advises of using the suggested `AddTrigger()` instead.

ItsShamed committed:
refactor(InputTrigger): move out of KCD

I love JetBrains Rider.

ItsShamed committed:
style: use Trigger initialisation

peppy committed:
Merge branch 'master' into gameplay/key-counter-abstraction

peppy committed:
Tidy up `Add` method logic

peppy committed:
Move `KeyCounter` components to `HUD` namespace

peppy committed:
Add basic xmldoc to `KeyCounter` classes

peppy committed:
Remove `NRT` disables in new classes

peppy commented:
I think this is pretty okay now.

@bdach requesting a final pass to make sure I haven't broken anything badly.

bdach committed:
Remove unused using directive

bdach committed:
Improve documentation of `InputTrigger`

bdach committed:
Reorder and add xmldoc to `KeyCounter` members

bdach commented on osu.Game/Screens/Play/HUD/KeyCounterDisplay.cs:
```diff
@@ -0,0 +1,113 @@
+            this.receptor = receptor;
+        }
+
+        public abstract void AddTrigger(InputTrigger trigger);
+
+        public void AddTriggerRange(IEnumerable<InputTrigger> triggers) => triggers.ForEach(AddTrigger);
+
+        public override void Add(KeyCounter counter)
+        {
+            if (!checkType(counter))
+                throw new InvalidOperationException($"{counter.GetType()} is not a supported counter type. (hint: you may want to use {nameof(AddTrigger)} instead.)");
+
+            base.Add(counter);
+            counter.IsCounting.BindTo(IsCounting);
+        }
```
I started to apply some final cleanups to this pull but this is where I get stuck. Why is this here still? As in, why is any subclass still allowed to arbitrarily add any `KeyCounter` if the type isn't matching?

Like this is so weird for me, because `KeyCounterDisplay` is apparently a `Container`, but then `DefaultKeyCounterDisplay` is setting `InternalChild` anyway, and thus is just nullifying the fact that its base class is a container.

I would like to see the process of creating the individual key counters and placing them in the key counter display's layout entirely encapsulated *within* the key counter display, because it shouldn't ever *make sense* to mix and match key counter implementations. It should be fundamentally impossible to misuse the component. This would be done by making `KeyCounter` a plain `CompositeDrawable`, with the expectation that implementors of `AddTrigger()` should in response create the `KeyCounter` they want and place it in the layout of the key counter display according to expectation.

peppy replied:
I did consider guarding against this further but wasn't sure it was worth the effort. Don't disagree with direction though.

peppy replied:
That said, I did initially try to refactor to a `CompositeDrawable` but it gets messy since `CompositeDrawable` doesn't support `Content` overrides.

As can be seen by https://github.com/ppy/osu/pull/22654/commits/44297a7d0a01e44f24d2860cdd4ce2584988e341.

So I'm not 100%, see what you think I guess.

ItsShamed committed:
refactor: make KCD a `CompositeDrawable`

ItsShamed committed:
docs: add XMLDoc for methods in KCD

ItsShamed committed:
refactor: make `Counters` return a `Container`

bdach commented on osu.Game/Screens/Play/HUD/KeyCounterDisplay.cs:
```diff
@@ -0,0 +1,111 @@
+        /// <summary>
+        /// The <see cref="KeyCounter"/>s contained in this <see cref="KeyCounterDisplay"/>.
+        /// </summary>
+        public abstract Container<KeyCounter> Counters { get; }
```
This being a `Container<KeyCounter>` again means that none of the last 3 commits actually changed anything in the state of affairs, because if it's exposed as a container, then you can add anything you like again through it.

Can we revert 5d15426c275084144d68e8790f0275d985c52269? I was honestly quite fine with the `IEnumerable<KeyCounter>`. I don't find it at all messy. The only consumers of this property are: test scenes, and `RulesetInputManager`. It's very contained.

ItsShamed replied:
The purpose of that commit was to fix this
https://github.com/ppy/osu/actions/runs/4360048022
Does that mean that I need to remove the checks for the counter flows visibility in `TestSceneHUDOverlay`? This is the only way I see to solve the failures (plus I think that this check will most likely be irrelevant as new variants get implemented)

bdach replied:
How does changing that from `IEnumerable` to `Container` fix test failures? That's an alarm bell if you ask me. Just changing the *returned type* fix it? Can you explain why?

ItsShamed replied:
Because of this??
https://github.com/ppy/osu/blob/8f6df5ea0f7f721c630fc8cad93bb3eef869d1d9/osu.Game.Tests/Visual/Gameplay/TestSceneHUDOverlay.cs#L48

ItsShamed replied:
Acutally the more I look into it, the more it feels weird to me

bdach replied:
Well this is where I would just fix the test to not use `.ChildrenOfType<>()`. But maybe it's just me.

ItsShamed replied:
I'll get into this later, I'm really not focused right now.

ItsShamed committed:
revert: make `counters` an `IEnumerable` again

As suggested by bdach as this would make the last two commits useless

Refs: 5d15426

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneSkinnableHUDOverlay.cs:
```diff
@@ -43,7 +42,7 @@ public partial class TestSceneSkinnableHUDOverlay : SkinnableTestScene
         // best way to check without exposing.
         private Drawable hideTarget => hudOverlay.KeyCounter;
-        private FillFlowContainer<KeyCounter> keyCounterFlow => hudOverlay.KeyCounter.ChildrenOfType<FillFlowContainer<KeyCounter>>().First();
+        private Drawable keyCounterFlow => (Drawable)hudOverlay.KeyCounter.Counters;
```
This is the dodgiest code I've seen in a *long time*. To the point my brain has trouble comprehending how it even compiles..

peppy committed:
Fix super-dodgy cast of `IEnumerable` to `Drawable`

peppy committed:
Use object initialisers and fix order of initialisation vs add

peppy committed:
Rename `Add`/`AddRange` methods as they are no longer conflicting with `Container`

peppy commented:
I've pushed remaining menial fixes. Kinda done with this PR at this point so going to go ahead with the merge. Any further changes can be applied in a new PR.

peppy committed:
Merge branch 'master' into gameplay/key-counter-abstraction

peppy committed:
Fix missing using statements in multiple test scenes

bdach committed:
Merge branch 'master' into gameplay/key-counter-abstraction

bdach committed:
Fix incorrect indent
